Allow for different versions of a question title depending on a conditional question

shahi645 opened the pull request:
### What is the context of this PR?
We have a use case in SR where the user answers a proxy question, e.g. "Who are you answering on behalf of?". This means that we want subsequent questions to be asked differently depending on the users answers. Previously this meant using routing rules and having a lot of excess schema.
This PR introduces a `titles` object to question, this can be used instead of a single `title` value. The titles object consists of a `default value` with is a single title, with optional extra `titles` that are accompanied with a `when` rule (`when` functionality already exists in routing rules, so we have re-used that code). When the question title is rendered each `when clause` is evaluated and if true returns its given `title`. If no `when clauses` evaluate true then the `default title` is returned and rendered.

**UPDATE**
We have added changes that now mean all getting of question titles are handled by our templating utils and template logic on selecting the question title is handled in our jinja filters.


***NOTE***
Currently we have a bug with repeating groups that are as follows: 
- The summary only renders the latest group instance of a repeating group. (existing bug)

- When evaluating a `when` clause in a repeating group greater than `group_instance` **0** it defaults to 0 `group_instance` for summary rendering, as summary has no knowledge of which group instance the question/answer has come from.

### How to review 
Run our test_schemas created and see if the changes are correct. 
Also does the code changes make sense as to where they are.

shahi645 commented:
please note travis build wont pass until https://github.com/ONSdigital/eq-survey-runner/pull/1605 is merged

nealedj commented on app/questionnaire/questionnaire_schema.py:
```diff
@@ -206,8 +206,27 @@ def _get_error_messages(self):
 
         return messages
 
+    def _get_answer_ids_in_question(self, question_id):
```
There's already a `get_answers_by_id_for_block` method. Could we follow the convention and rename this to `get_answer_ids_for_question`?

Following the convention you'd have

```
    def get_answers_by_id_for_question(self, question_id):
        question = self.get_question(question_id)
        if question:
            return {
                answer['id']: answer
                for answer in question['answers']
            }

        return {}

    def get_answer_ids_for_question(self, question_id):
        return list(self.get_answers_by_id_for_question(question_id).keys())

    def get_answers_for_question(self, question_id):
        return list(self.get_answers_by_id_for_question(question_id).values())
```

AndrewTorrance replied:
Done

nealedj commented on app/questionnaire/questionnaire_schema.py:
```diff
@@ -226,7 +246,30 @@ def _get_answer_dependencies(self):
+        for question in self.questions:
+            self._set_titles_dependencies_for_question(dependencies, question)
+
+    def _set_titles_dependencies_for_question(self, dependencies, question):
```
Lots of nesting here. Making it quite tricky to read.

Possible refactor:

```
        for title in question.get('titles', []):
            if 'when' not in title:
                continue

        when_clauses = title.get('when')

        for when_clause in when_clauses:
            for answer_id in answer_ids:
                when_id = when_clause.get('id')
                if answer_id not in dependencies[when_id]:
                    dependencies[when_id].append(answer_id)
```

AndrewTorrance replied:
Did a different implementation using list comprehensions.

nealedj commented on app/questionnaire/questionnaire_schema.py:
```diff
@@ -206,8 +206,27 @@ def _get_error_messages(self):
     def _get_answer_dependencies(self):
         dependencies = defaultdict(list)
+
+        self._get_answer_level_dependencies(dependencies)
```
Some of these methods are called `_set_xxx` and others `_get_xxx` but it's not clear why. As they're all mutating the `dependencies` object `set` might be more appropriate.

AndrewTorrance replied:
That one does not directly set the self.dependencies , it returns dependencies which other code assigns to the self.dependencies so I think its  valid as it does not off itself affect the object state.

nealedj replied:
The two `_get` methods and the `_set` method both mutate the `dependencies` argument that gets passed to them - I'm not why they employ different verbs.

Generally (subjectively) I think I'd expect a method start with `get` to not mutate any state. A `set` method might mutate the state of the object.

In this case perhaps `build` might be a better word to use?

AndrewTorrance replied:
I can see why its confusing . They are internal functions , none of which change the dependencies as stored on self. I have changed them all to gets as they get the dependencies . The actual set is done in the parse

nealedj commented on app/questionnaire/questionnaire_schema.py:
```diff
@@ -206,8 +206,27 @@ def _get_error_messages(self):
+        return answer_ids
+
     def _get_answer_dependencies(self):
         dependencies = defaultdict(list)
```
Consider making this a `defaultdict(set)` so that you can add to it without the need to check if a value is already present.

E.g.
```
>>> question_id, answer_id = 'foo', 'bar'
>>> dependencies = defaultdict(set)
>>> dependencies[question_id].add(answer_id)
>>> dependencies[question_id].add(answer_id)
{'foo': {'bar'}}
```

AndrewTorrance replied:
I like the idea , but the dependencies are used by existing code , so I'd rather leave as is for now.

nealedj commented on app/questionnaire/rules.py:
```diff
@@ -172,7 +172,10 @@ def evaluate_when_rules(when_rules, metadata, answer_store, group_instance):
         if 'id' in when_rule:
-            value = get_answer_store_value(when_rule['id'], answer_store, group_instance)
+            when_specified_group_instance = when_rule.get('group_instance')
+            instance = when_specified_group_instance if when_specified_group_instance is not None else group_instance
```
Could be reduced to:
`instance  = when_specified_group_instance or group_instance`

AndrewTorrance replied:
good catch

nealedj commented on app/templating/summary/question.py:
```diff
@@ -1,19 +1,37 @@
 
+    @classmethod
+    def get_title_from_titles(cls, metadata, answer_store, titles, group_instance):
+        when_rules = [(title['when'], title['value']) for title in titles if 'when' in title]
```
Nitpick - could be a generator:

```
when_rules = ((title['when'], title['value']) for title in titles if 'when' in title)
```

AndrewTorrance replied:
Its is nitpicky , but valid , changed

nealedj commented on app/templating/summary/question.py:
```diff
@@ -1,19 +1,37 @@
+    @classmethod
+    def get_title_from_titles(cls, metadata, answer_store, titles, group_instance):
+        when_rules = [(title['when'], title['value']) for title in titles if 'when' in title]
+        for when in when_rules:
```
Indexing a tuple rather than unpacking to variables.

Possibly more readable:

```
for when, value in when_rules:
    if evaluate_when_rules(when, metadata, answer_store, group_instance):
        return value
```

AndrewTorrance replied:
Changed , I was toying with a named tuple , but no need for 1 use simple unpacking is perfectly good

nealedj commented on app/templating/summary/question.py:
```diff
@@ -1,19 +1,37 @@
+
+    @classmethod
+    def get_question_title(cls, question_schema, answer_store, metadata):
+        if question_schema.get('title'):
```
What's the expected behaviour if the `title` is defined but == `""`?

Given that the templates are using the `is defined` jinja operator I'd say that for consistency we should treat empty strings as a defined title.

So this could become `if 'title' in question_schema:`

AndrewTorrance replied:
Its not that straightforward . The pre existing logic was if title =='' then use the value defined in the answer label as the title. So we changed the logic to
if schema[title] return schema[title]
if valid title in titles return that title
else return answer label for question 0 

I think should be revisited when looking at repeating groups .

nealedj commented on app/templating/summary/question.py:
```diff
@@ -1,19 +1,37 @@
+
+        titles = question_schema.get('titles')
+
+        return cls.get_title_from_titles(metadata, answer_store, titles, 0) if titles else question_schema['answers'][0]['label']
```
Suggest that this is opened onto separate lines for readability:

```
if 'titles' in question_schema:
    return cls.get_title_from_titles(metadata, answer_store, titles, 0)
else:
    return question_schema['answers'][0]['label']
```

AndrewTorrance replied:
mmm com si com sa , changed it without the else

nealedj commented on app/templating/summary/question.py:
```diff
@@ -1,19 +1,37 @@
+
+        titles = question_schema.get('titles')
+
+        return cls.get_title_from_titles(metadata, answer_store, titles, 0) if titles else question_schema['answers'][0]['label']
```
We're explicitly setting the `group_instance` to `0` here but without explanation.

We know it's because summaries don't yet support repeating groups - could we put a comment to explain?

AndrewTorrance replied:
Added

nealedj commented on data/en/test_different_question_titles.json:
```diff
@@ -0,0 +1,204 @@
+                        "title": "Who are you answering on behalf of?",
+                        "guidance": {
+                            "content": [{
+                                "description": "The answer you choose will have an affect on question titles in next question"
```
effect ;)

AndrewTorrance replied:
:-)

nealedj commented on scripts/test_schemas.sh:
```diff
@@ -6,8 +6,8 @@ if [ "$1" == "--local" ] || [ "$2" == "--local" ]; then
 if [ "$run_docker" == true ]; then
-    docker pull onsdigital/eq-schema-validator
-    validator="$(docker run -d -p 5001:5000 onsdigital/eq-schema-validator)"
+    docker pull onsdigital/eq-schema-validator:eq-1990-different-versions-of-question-titles
```
Note to remove this before merge.

AndrewTorrance replied:
Added comment to script. PR also has do not merge on it

nealedj commented on tests/app/forms/test_questionnaire_form.py:
```diff
@@ -869,6 +869,40 @@ def test_multi_calculation(self):
             self.assertEqual(form.question_errors['breakdown-question'],
                              schema.error_messages['TOTAL_SUM_NOT_EQUALS'] % dict(total='10'))
 
+    def test_generate_form_with_titles(self):
```
I don't understand what this is testing.

nealedj commented on app/templating/view_context.py:
```diff
@@ -53,13 +53,18 @@ def build_view_context_for_question(current_location, variables, rendered_block,
             'other_answer': {},
-            'fields': {}
+            'fields': {},
+            'question_titles': {}
```
I don't think the `question_titles` should be nested inside the `form` object. They don't belong to it.

In the absence of a proper view model I'd suggest patching the title directly onto the `question` objects within each `rendered_block`.

```
for question in rendered_block.get('questions'):
    question['question_title`] = get_the_title()
```

Then the templates wouldn't need any logic for working out where to get the title from.

AndrewTorrance replied:
We raised something similar as a possibility at the beginning but we shied away from that after sharing the ideas because of the need to know if the question was inside a repeating group / summary . We can do it but it would mean a significant rewrite .

nealedj commented on app/templating/view_context.py:
```diff
@@ -5,10 +5,10 @@
 from app.helpers.form_helper import get_form_for_location
 from app.templating.summary_context import build_summary_rendering_context
 from app.templating.template_renderer import renderer
+from app.templating.summary.question import Question
```
This is being imported so that the `get_title_from_titles` function can be called. I think that implies that the function is in the wrong place because it isn't specific to the summary.

Consider adding a `utils` module to the `templating` package for now and add the function there instead.

AndrewTorrance replied:
Your spot on. Adding it to view context caused circular dependencies . Moving out to a utils file

nealedj reviewed (changes requested):
Well done - loads of good stuff here. Lots of little comments but nothing fundamental. The main thing that I think could be improved is where the titles are rendered into the view context.

General point - could the test schemas be named such that they're grouped together? Might be useful for these bigger features that require multiple schemas.

E.g. test_titles.json, test_titles_within_repeating_blocks.json, test_titles_radio_checkbox

shahi645 commented:
@nealedj ready for re-review

davecharles commented on app/forms/questionnaire_form.py:
```diff
@@ -298,12 +299,12 @@ def map_child_option_errors(errors, answer_json):
 
 
-def generate_form(schema, block_json, data, answer_store, metadata):
+def generate_form(schema, block_json, data, answer_store, metadata, group_instacne):
```
Bug in waiting... `group_instacne`

davecharles commented on app/jinja_filters.py:
```diff
@@ -226,6 +226,47 @@ def max_value(a, b):
 
+@contextfunction
+@blueprint.app_template_filter()
+def get_question_title(context, question_id):
```
Existing filters have adopted the convention of `format_` - could we call theses `format_question_title` etc?

shahi645 replied:
i think in this instance its better as a get, we're not mutating anything here we're just getting a value based on the context

davecharles commented on app/jinja_filters.py:
```diff
@@ -226,6 +226,47 @@ def max_value(a, b):
+
+
+@blueprint.app_context_processor
+def get_question_title_check():
```
Maybe `question_title_processor` for consistency?

shahi645 replied:
yeah can do, the others have check but you're right processor probably sounds better

davecharles commented on app/templating/utils.py:
```diff
@@ -0,0 +1,22 @@
+from app.questionnaire.rules import evaluate_when_rules
```
Don't understand why the separate utils module.  It's very 'titles' centric so not really utils. Less cognitive load for the reader if it's just in `view_context.py` - unless there's a reason I'm missing (e.g. avoiding circ import).

shahi645 replied:
its called in several places, so we agreed in the utils would be best served

nealedj replied:
Ideally this logic would form part of a `Question` view model of some type. But we don't have that concept and it's quite a big one to introduce.

Separate module is because it's used by both the context builder module and the summary view models.

And yep, there would have been a circular import if it was defined in `view_context.py` since that imports the summary view model which depends on this function.

davecharles commented on app/views/questionnaire.py:
```diff
@@ -703,14 +703,19 @@ def _get_front_end_navigation(answer_store, current_location, metadata, routing_
 
-def get_page_title_for_location(schema, current_location):
+def _get_question_title_for_page(first_question, context):
+    return first_question['title'] if first_question.get('title') is not None else context['question_titles'][first_question['id']]
```
I think this is a bit more readable:
```python
if first_question.get('title') is not None:
    return first_question['title']
return context['question_titles'][first_question['id']]
```
Do you explicitly want to test for None and not empty str as well?

shahi645 replied:
its tested in `test_questionnaire`

davecharles commented on tests/app/forms/test_household_composition.py:
```diff
@@ -20,7 +20,7 @@ def _error_exists(answer_id, msg, mapped_errors):
     def test_get_name_form_generates_name_form(self):
         with self.test_request_context():
-            form = generate_household_composition_form(self.schema, self.block_json, {})
+            form = generate_household_composition_form(self.schema, self.block_json, {}, metadata={}, group_instance=0)
```
The function def only has positional args so technically you don't need to name `metadata` and `group_instance` keywords. Not a big issue but I'm wondering if it says something about `generate_household_composition_form` args - do we need a `*` arg list delimiter for example?

shahi645 replied:
i just named them for clarity, i guess thats a better standard but not sure myself. Raises food for thought though, all tutorials i have watched have mentioned to name args regardless

nealedj replied:
As a rule I never name positional args and always name kwargs. TBH I've never questioned it but all python devs I've worked with have done the same.

davecharles commented on tests/functional/spec/features/conditional_question_title/conditional_titles_in_repeating_groups.spec.js:
```diff
@@ -0,0 +1,40 @@
+const helpers = require('../../../helpers');
+
+describe('@watch Feature: Use of Titles in Repeating blocks', function() {
```
Need to remove this @watch

AndrewTorrance replied:
Done

nealedj commented on app/views/questionnaire.py:
```diff
@@ -714,14 +714,21 @@ def _get_front_end_navigation(answer_store, current_location, metadata, routing_
 
-def get_page_title_for_location(schema, current_location):
+def _get_question_title_for_page(first_question, context):
+    if first_question.get('title') is not None:
```
This might make more sense as `if 'title' in first_question`. My reasoning being that as a rule if `title` is defined on a schema then it should take precedence over `titles` - even if `title` is an empty string.

All this check does is adds implicit behaviour if somebody were to define `"title": null` (which validator wouldn't allow anyway).

nealedj reviewed (approved):
Really good work chaps. I've left one minor comment to consider but I'm not blocking it based on that.

dwyeradam commented:
Apparently I'm the Simon Cowell of eQ reviews and the designated merger here. Just doing a pass of the code and i'll merge it when i'm done.

dwyeradam commented on app/questionnaire/questionnaire_schema.py:
```diff
@@ -206,9 +206,33 @@ def _get_error_messages(self):
+        dependencies.update(self._get_question_level_dependencies())
+
+        dependencies.update(self._get_titles_dependencies_for_all_questions())
+
```
This is a type of question level dependency so rather than loop through the questions twice it's worth calling `_get_titles_dependencies_for_question(...)` within `_get_question_level_dependencies()`. 
It may also be worth moving the existing code in `_get_question_level_dependencies()` into it's own method, something like ``_get_calculations_dependencies_for_question(...)``

dwyeradam commented on app/questionnaire/questionnaire_schema.py:
```diff
@@ -217,17 +241,44 @@ def _get_answer_dependencies(self):
+            self._get_titles_dependencies_for_question(dependencies, question)
         return dependencies
 
+    def _get_titles_dependencies_for_question(self, dependencies, question):
```
I'm not a fan of mutating parameters. We'd have to test if we can safely merge the dictionary though if keys already exist. We may already have an issue here.

nealedj commented on app/questionnaire/dependencies.py:
```diff
@@ -0,0 +1,39 @@
+    i.e  a['question_id']
+    indexing is read only
+    """
+    def __init__(self):
```
Instead of holding a defaultdict internally maybe consider extending a `defaultdict` for this class.

```
class Dependencies(defaultdict):
  def __init__(self):
    super().__init__(set)

  def update(*args, **kwargs):    
    for other in args:  
      self._merge(other)
    if kwargs:
      self._merge(kwargs)

  def _merge(self, other):
    for key, dependencies in other.items():
      self[key] |= dependencies
```

(untested)

nealedj commented on app/questionnaire/dependencies.py:
```diff
@@ -0,0 +1,39 @@
+    set used to prevent duplicates
+
+    To add a dependency either call add with a single dependency
+    or a+=b where both a and b are Dependencies objects.
```
I'm not sure it's a good idea to implement the `+` operator for the `Dependencies` object when the `Dependencies` object itself is largely acting as a dict. Dicts can't be added together, that's what `update()` is for, so I'd suggest implementing that instead.

nealedj commented on app/questionnaire/dependency_builder.py:
```diff
@@ -0,0 +1,74 @@
+from app.questionnaire.dependencies import Dependencies
+
+
+class DependenciesBuilder:
```
Since the dependencies state is held and managed by the `Dependencies` type I don't think there's much value in this being a class itself.

In a static language this would be a static class - it holds no state. In python I'd suggest that it's more pythonic to define the methods as functions on the module instead.

shahi645 committed:
Allow for different version of question titles

feature includes:

- titles object can be added to question schema

- titles can have multiple versions of a question title

- title chosen based on conditional answer given

- backward compatible, so title can still be used
